File: Algorithm/과제 8/music.cpp
#include <iostream>
#include <fstream>
#include <tuple>
#include <string>
using namespace std;

tuple<string, string, int, int, int> cal(string first, string second, int match, int mis, int gap);
int max(int a, int b, int c, int d);

int main(void) {
	//입력을 받는 부분
	ifstream in("music.inp");
	ofstream out("music.out");

	int match, mis, gap, best = 0, f_best, s_best;
	string str, first, second;
	in >> match >> mis >> gap;
	getline(in, str);
	getline(in, str);

	//계산하는 부분
	for(int i = 1; i < str.length()-1; i++) {
		tuple<string, string, int, int, int> ret = cal(str.substr(0, i), str.substr(i), match, mis, gap);

		if(get<2>(ret) > best) {
			first = get<0>(ret);
			second = get<1>(ret);
			best = get<2>(ret);
			f_best = get<3>(ret);
			s_best = get<4>(ret) + i;
		}
		else if(get<2>(ret) == best ){
			if(f_best > get<3>(ret))
			{
				first = get<0>(ret);
				second = get<1>(ret);
				best = get<2>(ret);
				f_best = get<3>(ret);
				s_best = get<4>(ret) + i;
			}
			else if(f_best == get<3>(ret) && s_best > get<4>(ret) + i)
			{
				first = get<0>(ret);
				second = get<1>(ret);
				best = get<2>(ret);
				f_best = get<3>(ret);
				s_best = get<4>(ret) + i;
			}
		}
	}

	//출력하는 부분
	string result = first + "\n" + second + "\n" + to_string(best);
	out.write(result.c_str(), result.size());

	in.close();
	out.close();
	return 0;
}

tuple<string, string, int, int, int> cal(string first, string second, int match, int mis, int gap) {
	int best = 0;
	int bLoc[2] = {1, 1};

	int f_len = first.length() + 1;
	int s_len = second.length() + 1;

	int** arr = new int* [f_len];
	for(int i = 0; i < f_len; i++)
		arr[i] = new int[s_len] {0};

	for(int i = 1; i < f_len; i++) {
		for(int j = 1; j < s_len; j++) {
			int tmp = arr[i-1][j-1];
			if(first[i-1] == second[j-1])
				tmp += match;
			else
				tmp += mis;

			arr[i][j] = max(tmp, arr[i][j-1] + gap, arr[i-1][j] + gap, 0);

			if(arr[i][j] > best) {
				best = arr[i][j];
				bLoc[0] = i;
				bLoc[1] = j;
			}
		}
	}

	int start[2] = {bLoc[0], bLoc[1]};

	while(true) {
		if(start[0] == 0 && start[1] == 0) {
			break;
		}
		else if(start[0] == 0) {
			if(arr[start[0]][start[1]-1] != 0)
				start[1] -= 1;
			else
				break;
		}
		else if(start[1] == 0) {
			if(arr[start[0]-1][start[1]] != 0)
				start[0] -= 1;
			else
				break;
		}
		else{
			if (arr[start[0]-1][start[1]-1] != 0) {
        		start[0] -= 1;
        		start[1] -= 1;
    		}
    		else if (arr[start[0]-1][start[1]] != 0)
        		start[0] -= 1;
    		else if (arr[start[0]][start[1]-1] != 0)
        		start[1] -= 1;
    		else
            	break;
		}
	}

	for(int i = 0; i < f_len; i++)
        delete [] arr[i];

    return tuple<string, string, int, int, int>(first.substr(start[0]-1, bLoc[0]+1-start[0]), second.substr(start[1]-1, bLoc[1]+1-start[1]), best, start[0], start[1]);
}

int max(int a, int b, int c, int d)
{
    if(a >= b && a >= c && a >= d)
    	return a;
    else if(b >= c && b >= d)
    	return b;
    else if(c >= d)
    	return c;
    else
    	return d;
}

File: Algorithm/과제 8/music.py
def read_data(): #데이터 입력
    f = open("music.inp", "r")
    match, mis, gap = map(int, f.readline().split())
    string = f.readline().rstrip()
    return string, match, mis, gap

def print_data(first, second, best): #데이터 출력
    f = open("music.out", "w")
    tmp = first + "\n" + second + "\n" + str(best)
    f.write(tmp)
    f.close()

def process(string, match, mis, gap):
    sub_str = []

    f = ""
    s = ""
    b = 0
    
    for i in range(1, len(string)-2):
        sub_str.append([string[:i],string[i:]])

    for i in sub_str:
        tmp_f, tmp_s, tmp_b = cal(i[0], i[1], match, mis, gap)
        if tmp_b > b:
            f = tmp_f
            s = tmp_s
            b = tmp_b
    
    return f, s, b

def cal(first, second, match, mis, gap):
    best = 0
    bLoc = [0,0]

    f_len = len(first) + 1
    s_len = len(second) + 1

    arr = [[0 for _ in range(f_len)]for _ in range(s_len)]

    for i in range(1, s_len):
        for j in range(1, f_len):
            tmp = arr[i-1][j-1] + [match, mis][first[j-1] != second[i-1]]
            arr[i][j] = max(tmp, arr[i][j-1] + gap, arr[i-1][j] + gap, 0)
            if arr[i][j] > best:
                best = arr[i][j]
                bLoc = [j-1, i-1]

    start = [bLoc[0], bLoc[1]]
    while True:
        if arr[start[1]-1][start[0]-1] != 0:
            start[0] -= 1
            start[1] -= 1
        elif arr[start[1]-1][start[0]] != 0:
            start[1] -= 1
        elif arr[start[1]][start[0]-1] != 0:
            start[0] -= 1
        else:
            break

    return (first[start[0]-1:bLoc[0]+1], second[start[1]-1:bLoc[1]+1], best)

string, match, mis, gap = read_data()
f, s, b = process(string, match, mis, gap)
print_data(f, s, b)
